lib/stores/S3Store.js

```diff
@@ -438,138 +500,141 @@ class S3Store extends DataStore {
-                                            upload_length: metadata.file.upload_length,
-                                        },
-                                        isFinished: true,
-                                    };
-                                })
-                                .catch((headError) => {
-                                    throw headError;
-                                });
-                        });
-                })
-                .then((data) => {
-                    if (data.isFinished) {
-                        return resolve(data.data);
-                    }
-                    // if no parts are found, offset is 0
-                    if (data.parts.length === 0) {
-                        return resolve({
-                            size: 0,
-                            upload_length: data.metadata.file.upload_length,
-                        });
-                    }
-
-                    const offset = data.parts.reduce((a, b) => {
-                        a += parseInt(b.Size, 10);
+        return this._getMetadata(file_id)
+            .then((metadata) => {
+                return this._retrieveParts(file_id)
+                    .then((parts) => ({ parts, metadata }))
+                    .catch((err) => {
+                        // getOffset is called from both the HEAD request and the PATCH request
+                        // (via the _processUpload method). The HEAD request calls it (implicitly)
+                        // with with_parts: false, the _processUpload method calls it (explicitly)
+                        // with with_parts: true. The _processUpload method expects upload parts to be returned,
+                        // which would not work when the upload is completed, since there are no parts anymore.
+                        // TODO: refactor to separate these concerns.
+                        // TODO: write a test for this case.
+                        if (with_parts === true || err.code !== 'NoSuchUpload') {
+                            throw err;
+                        }
+                        return this.client
+                            .headObject({ Bucket: this.bucket_name, Key: file_id })
+                            .promise()
+                            .then(() => {
+                                return {
+                                    data: {
+                                        ...this.cache[file_id].file,
+                                        size: metadata.file.upload_length,
+                                        upload_length: metadata.file.upload_length,
+                                    },
+                                    isFinished: true,
+                                };
+                            })
+                            .catch((headError) => {
+                                throw headError;
+                            });
+                    });
```

Comment:
@mitjap this whole section was implemented after your initial version of this PR in #207. However, since you're rewrite this is no longer conditionally called with `with_parts` so we can remove the argument and clean this up

Reply:
Solved in https://github.com/tus/tus-node-server/pull/245/commits/28ced0409b94cf4495a282da90f3d4f13298f2d5